8347067: Load certificates without explicit trust settings in KeyChainStore

timja opened the pull request:
<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

## The change

Without this change intermediate certificates that don't have explicit trust settings are ignored not added to the truststore.



## Reproducer

See https://github.com/timja/openjdk-intermediate-ca-reproducer

Without this change the reproducer fails, and with this change it succeeds.

## Example failing architecture

Root CA -> Intermediate 1 -> Intermediate 2 -> Leaf

Where:
* All certs are in admin domain kSecTrustSettingsDomainAdmin
* Root CA is marked as always trust
* Intermediate 1 and 2 are Unspecified

Previously Root CA would be found but intermediate 1 and 2 would be skipped when verifying trust settings.

## Background reading

### Rust
see also Rust Lib that is used throughout Rust ecosystem for this: 
https://github.com/rustls/rustls-native-certs/blob/efe7b1d77bf6080851486535664d1dc7ef0dea68/src/macos.rs#L39-L58

e.g. in Deno `https://github.com/denoland/deno/pull/11491` where I've verified it is correctly implemented and works in my setup

## Python

I also looked at the Python implementation for inspiration as well (which also works on my system): https://github.com/sethmlarson/truststore/blob/main/src/truststore/_macos.py

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8347067](https://bugs.openjdk.org/browse/JDK-8347067): Load certificates without explicit trust settings in KeyChainStore (**Enhancement** - P4)


### Reviewers
 * [Alexey Bakhtin](https://openjdk.org/census#abakhtin) (@alexeybakhtin - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22911/head:pull/22911` \
`$ git checkout pull/22911`

Update a local copy of the PR: \
`$ git checkout pull/22911` \
`$ git pull https://git.openjdk.org/jdk.git pull/22911/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22911`

View PR using the GUI difftool: \
`$ git pr show -t 22911`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22911.diff">https://git.openjdk.org/jdk/pull/22911.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22911#issuecomment-2613394026)
</details>

timja commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -411,15 +411,16 @@ static bool loadTrustSettings(JNIEnv *env,
     CFArrayRef trustSettings;
-    // Load trustSettings into inputTrust
-    if (SecTrustSettingsCopyTrustSettings(certRef, domain, &trustSettings) == errSecSuccess && trustSettings != NULL) {
+    if (*inputTrust == NULL) {
```
moved so that the empty `ArrayList` is always created

timja commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -492,11 +493,6 @@ static void addCertificatesToKeystore(JNIEnv *env, jobject keyStore,
             }
 
-            // Only add certificates with trust settings
-            if (inputTrust == NULL) {
```
From what I can tell non root certificates that do not have explicit trust settings do not show up in `SecTrustSettingsCopyTrustSettings`, docs appear to be https://developer.apple.com/documentation/security/sectrustsettingscopytrustsettings(_:_:_:) but not very clear.

----

I need to test that the certificate is still chained to a root and not trusted as a root.

timja replied:
Ok this isn't working properly 😢 

1. ⛔ Fails: Marking the certificate as OS default (which for CA certs is trust: false) - with an intermediate (I haven't tried without yet)
2. ⛔ Fails: Removing the root but leaving the intermediate

timja replied:
Interesting for root certificate `SecTrustSettingsCopyTrustSettings` returns:

* -25300 (not found) when trust policy is `Use System Defaults`
* 0 and a `kSecTrustSettingsResult` value of 3 when set to Never Trust
* 0 and a `kSecTrustSettingsResult` value of 1 when set to Always Trust

timja replied:
With https://github.com/openjdk/jdk/pull/22911/commits/5102dade13f44dedd887920c407158e7d189947b

Case 2. works again.

(i.e. the basic case which previously worked with a self-signed root and no intermediate)

Case 1 and 3 are still failing, I'll have a think on Monday but may need a pointer

timja replied:
With https://github.com/openjdk/jdk/pull/22911/commits/0052cd0380b4949b9af689eae660cf3defa5e7d0 all cases are working

alexeybakhtin commented:
I'm afraid but it looks like this implementation violates the https://developer.apple.com/documentation/security/sectrustsettingscopytrustsettings(_:_:_:) specification.

It always creates an empty input trust and fills it with the provided non-null trustSettings. So, the certificate always has at least empty trustSettings. However, according to this specification, two different scenarios should be considered:
1. Empty trustSettings means that the certificate is trusted if it is self-signed.
2. Null trustSettings means that the certificate must be verified before adding to the trust store.

The current implementation adds Root certificate to the TrustStore because it has "Always Trust" settings.
Implementation does not add the intermediate certificate to the TrustStore because it has "System Defaults" settings ( SecTrustSettingsCopyTrustSettings returns null trustSettings), and the current implementation only adds certificates with trust settings.

I think, in this particular case, we need two iterations to add certificates into the trust store. The first iteration will add certificates with non-null trust settings, and the second iteration should verify and add certificates with null trust settings.

timja commented:
> I think, in this particular case, we need two iterations to add certificates into the trust store. The first iteration will add certificates with non-null trust settings, and the second iteration should verify and add certificates with null trust settings.

Thanks for the feedback it was very helpful, I had missed the bottom note on https://developer.apple.com/documentation/security/sectrustsettingscopytrustsettings(_:_:_:) before this.

I've implemented the recommendation based on the docs in https://github.com/openjdk/jdk/pull/22911/commits/0052cd0380b4949b9af689eae660cf3defa5e7d0.

All my test cases are now passing.

I've added a second intermediate CA to my test setup as well although it only uses 1 by default:
https://github.com/timja/openjdk-intermediate-ca-reproducer?rgh-link-date=2025-01-03T11%3A28%3A01Z

I've tested by revoking trust along each part of the chain and its behaving correctly now.

alexeybakhtin commented on src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:
```diff
@@ -458,6 +458,31 @@ static bool createTrustedCertEntry(JNIEnv *env,  jobject keyStore,
+   if (policy) {
+       CFRelease(policy);
+   }
+   if (secTrust) {
```
I think subjCerts should be released here too

alexeybakhtin commented:
Thank you for this patch. It looks correct now (see my comment about subjCerts above)

timja commented:
> Thank you for this patch. It looks correct now (see my comment about subjCerts above)

Thanks, will look into that

> Is it possible to add jtreg test for this scenario? 

I'll look at that ~tomorrow

> Also, You'll need a jbs issue to submit this PR

Would it be possible for you to do it on my behalf please? I don't have access

timja commented:
> Is it possible to add jtreg test for this scenario?

I've done some research.

I _think_ it would only be possible with manual intervention to run it.
The certificates could be generated with a script, similar to the existing https://github.com/openjdk/jdk/blob/master/test/jdk/sun/security/x509/DNSName/certs/generate-certs.sh and then checked in.

The certificates could be added to the truststore using `/usr/bin/security add-trusted-cert`, like in https://github.com/JetBrains/jvm-native-trusted-roots/blob/trunk/src/test/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtilTest.java#L114-L120

but marking the root certificate as trusted would need the user to confirm an OS prompt, https://github.com/JetBrains/jvm-native-trusted-roots#testing, i.e. I need to approve via Touch ID when I make changes to a certs trust level.

Does that add value to add a test so someone could run it manually?

alexeybakhtin commented:
/issue JDK-8347067

timja commented:
/issue JDK-8347067

timja commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,170 @@
+import org.junit.jupiter.api.Test;
+
+/*
+ * @test
```
@alexeybakhtin quick question on how this should be marked as manual.

I see all tests in https://github.com/openjdk/jdk/blob/master/test/jdk/TEST.groups#L256-L259 are manual ones.
Is this test automatically included in that?

Or should it be added here?
https://github.com/openjdk/jdk/blob/master/test/jdk/TEST.groups#L657

alexeybakhtin replied:
The test should be marked as `@run junit/manual ` and added to the `jdk_security_manual_interactive` part of the TEST.groups

timja replied:
Any idea how I can run the test after making those changes?

The test just gets skipped with:

```
CONF=release make test TEST=test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java
```

timja replied:
I was able to run it by setting up `jtreg` and avoiding the `make test` task.

```
jtreg.sh -manual test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java
```

timja commented:
Attaching logs are requested on mailing list,

[debug-all.txt](https://github.com/user-attachments/files/18394763/debug-all.txt)

[cert-path.txt](https://github.com/user-attachments/files/18394767/cert-path.txt)

timja commented:
(OCA has been signed by my employer, so should be able to move forward soon)

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,169 @@
+
+    private static void assertThat(boolean expected, String message, List<X509Certificate> certificates) {
+        if (!expected) {
+            throw new AssertionError(message + ", subjects: " + getSubjects(certificates));
```
I do not like printing all KeyChain certificates on the failure. It could be sensitive information.

If you do not collect all certificates, the test could be simplified - without Stream API

timja replied:
Without this its harder to debug what went wrong, this tells you what subjects are found, its not the certificate itself, just e.g. 

```
java.lang.AssertionError: Non trusted CA not found CN=Non Trusted Example CA,O=Example,C=US, subjects: [ CN=StartCom Certification Authority,OU=Secure Digital Certificate Signing,O=StartCom Ltd.,C=IL, CN=TIMJA-INTERMEDIATE,O=TIMJA,ST=ES,C=UK, CN=TIMJA-INTERMEDIATE-2,O=TIMJA,ST=ES,C=UK, CN=TIMJA-ROOT,O=TIMJA,ST=ES]
```

alexeybakhtin replied:
I understand the reason for the test development, but I am still concerned about printing this information.
Is it possible it will be disabled by default but could be enabled from the cmdline or changing the source code?
`
    static boolean verbose = false; // avoid too verbose output
`

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,169 @@
+ * @requires os.family == "mac"
+ * @summary Check whether loading of certificates from macOS Keychain correctly
+ *          loads intermediate CA certificates
+ * @run junit/manual CheckMacOSKeyChainIntermediateCATrust
```
I think you can add negative tests also. E.g. add Root CA without trust settings

alexeybakhtin replied:
Thank you for adding the test

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,169 @@
+    private static void executeProcess(List<String> params) {
+        System.out.println("Command line: " + params);
+        try {
+            int exitStatus = ProcessTools.executeProcess(params.toArray(new String[0])).getExitValue();
```
shouldHaveExitValue(0) can be used

seanjmullan commented:
/reviewers 2

seanjmullan commented:
This change is significant and should be reviewed by at least 2 Reviewers.

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,192 @@
+        assertThat(containsSubjectName(certificates, intermediateCASubjectName), "Intermediate CA not found " + intermediateCASubjectName, certificates);
+
+        String nonTrustedCASubjectName = "CN=Non Trusted Example CA,O=Example,C=US";
+        assertThat(not(containsSubjectName(certificates, nonTrustedCASubjectName)), "Non trusted CA found " + nonTrustedCASubjectName, certificates);
```
Could you please add one more test, if you don't mind? 

NonTrustedIntermediateCA is issued by nonTrustedCA:
```
openssl genrsa -out non-trusted-intermediate.key 2048
openssl req -new -sha256 -nodes -key non-trusted-intermediate.key  \
  -subj "/C=US/O=Example/CN=Non Trusted Example Intermediate CA" -out non-trusted-intermediate-ca.csr

openssl x509 -req \
 -extensions v3_ca \
 -extfile openssl.cnf \
 -in non-trusted-intermediate-ca.csr \
 -CA non-trusted-test-ca.pem \
 -CAkey non-trusted-root.key \
 -CAcreateserial \
 -out non-trusted-intermediate-ca.pem \
 -days 3650 \
 -sha256
```
In this case, we'll cover all basic scenarios.

alexeybakhtin commented on test/jdk/java/security/KeyStore/CheckMacOSKeyChainIntermediateCATrust.java:
```diff
@@ -0,0 +1,204 @@
+        if (!expected) {
+            String errMessage = message;
+            if (verbose) {
+                errMessage += ", subjects: " + getSubjects(certificates);
```
By default, subjects and elements in the subjects are separated by a comma. Can you change the separator between subjects?

alexeybakhtin reviewed (approved):
Thank you for the fixes. Looks good to me

timja commented:
Possible to get a second reviewer? potentially @wangweij?

Unless @seanjmullan you could suggest someone?

timja commented:
Hi @seanjmullan can you recommend anyone else for review please?

seanjmullan commented:
I am dubious that this is the right thing to do. There is a distinct difference between a certificate that is trusted and one that requires additional validation to determine if it is trusted. Blindly trusting self-signed certificates is very dangeorous, and unless I am mistaken, this seems to be what this code may do. I do not think lumping both of these certificates into "trusted" certificates is the right thing to do.

It may mean that we need to enhance the `KeyStore` API to be able to store arbitrary certificates. Or you could consider storing these additional certificates in a `CertStore`.

I am also not sure the use case is common. Typically a web server will send its entire chain, and not just the server certificate.

Certainly open to hearing alternate opinions if I have overlooked something.

timja commented:
> I am dubious that this is the right thing to do. There is a distinct difference between a certificate that is trusted and one that requires additional validation to determine if it is trusted. Blindly trusting self-signed certificates is very dangeorous, and unless I am mistaken, this seems to be what this code may do. I do not think lumping both of these certificates into "trusted" certificates is the right thing to do.

There is no 'blind' trust in this pull request. Any 'non trusted' certificate is validated via the OS APIs, once it has been validated by the OS then it is a trusted certificate it just needed its chain verified to the root.

https://github.com/openjdk/jdk/pull/22911/files#diff-b0b2f959f0fa18e568c0688bd233882283bca94f4ff6dce886a21510ff82d64dR475

https://github.com/timja/jdk/blob/59ceebc4acf2c39c86c45b1aa6746bf619417cea/src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m#L475

> It may mean that we need to enhance the `KeyStore` API to be able to store arbitrary certificates. Or you could consider storing these additional certificates in a `CertStore`.

The `KeyStore` can store them arbitrarily right now can't it but is it that currently anything in there is 'trusted'?

> 
> I am also not sure the use case is common. Typically a web server will send its entire chain, and not just the server certificate.

I don't think its very common either but I think its becoming more common with ZTNA providers (Zero Trust Network Access) being deployed in a lot of organisations for compliance and security reasons. 

When these are third party providers they often won't have your root cert and they may or may not provide the complete chain (I don't know if its configurable to do that) my organisation's one certainly doesn't provide the whole chain and there are other bug reports out there from other organisations too. 


> Certainly open to hearing alternate opinions if I have overlooked something.

An alternative option would be a truststore or some other mechanism that just does [SecTrustEvaluateWithError](https://developer.apple.com/documentation/security/sectrustevaluatewitherror(_:_:)?language=objc) without any of the complexity of providing the actual certificates but just checks the trust for the presented chain. Not sure how complex that would be.

alexeybakhtin commented:
> > > I am dubious that this is the right thing to do. There is a distinct difference between a certificate that is trusted and one that requires additional validation to determine if it is trusted. Blindly trusting self-signed certificates is very dangeorous, and unless I am mistaken, this seems to be what this code may do. I do not think lumping both of these certificates into "trusted" certificates is the right thing to do.
> > 
> > 
> > There is no 'blind' trust in this pull request. Any 'non trusted' certificate is validated via the OS APIs, once it has been validated by the OS then it is a trusted certificate it just needed its chain verified to the root.
> > 
> 
> That helps, but just because it successfully validated once does not mean it should be treated as a trusted certificate (essentially as a trust anchor). It may later be revoked/expired, its key or signature algorithm may weaken or it may fail validation for some other reason, or CA certificates further up the chain may change and affect the trustworthiness of this certificate.
 
Exactly the same happens with other certificates with trust settings. They are loaded on start and not verified for trust settings later. They are still trusted, If the Trust Settings are updated manually in the native KeyChainStore. Also, it affects KeychainStore only and does not affect KeychainStore-ROOT certificates, as it is static.

ecki commented:
Isnt that the Same for Windows? Not all Stores are explicitely trusted, some contain intermediate certificates. It would certainly be helpful to reflect trust in Keystore API (if only to not Trust the ones which are not marked as such). For example CurrentUser\Root or LocalMachine\CA are such Locations which could ne marking as trusted/not trusted respektively.

seanjmullan commented:
We need to be really careful here. With this fix we are deciding at runtime that these intermediate certificates should be treated as `KeyStore.TrustedCertificateEntry` objects just because they validated ok, and without any interaction with the user or application.

Also, the JDK does not rely on certificate chain validation from the OS. The JDK has its own PKIX `CertPathValidator` implementation and has its own restrictions on weak algorithms, etc which is a key part of establishing trust in certificates used in TLS, signed JARs, etc. You are now delegating this to MacOS as a mostly invisible change which brings in a new set of security concerns which may make it less secure or at a minimum requires resources to ensure this code is properly reviewed, audited, etc.
